django/contrib/contenttypes/fields.py

```diff
@@ -500,15 +500,35 @@ def get_prefetch_queryset(self, instances, queryset=None):
                     False,
                     self.prefetch_cache_name)
 
-        def add(self, *objs):
+        def add(self, *objs, **kwargs):
+            bulk = kwargs.pop('bulk', True)
             db = router.db_for_write(self.model, instance=self.instance)
-            with transaction.atomic(using=db, savepoint=False):
+
+            def check_and_update_obj(obj):
+                if not isinstance(obj, self.model):
+                    raise TypeError("'%s' instance expected, got %r" % (
+                        self.model._meta.object_name, obj
+                    ))
+                setattr(obj, self.content_type_field_name, self.content_type)
+                setattr(obj, self.object_id_field_name, self.pk_val)
+
+            if bulk:
+                ids = []
                 for obj in objs:
-                    if not isinstance(obj, self.model):
-                        raise TypeError("'%s' instance expected" % self.model._meta.object_name)
-                    setattr(obj, self.content_type_field_name, self.content_type)
-                    setattr(obj, self.object_id_field_name, self.pk_val)
-                    obj.save()
+                    if obj._state.adding or obj._state.db != db:
+                        raise ValueError("%r instance isn't saved, but must be." % obj)
```

Comment:
This exception message is different from that in `related.py` though the logic/intention surrounding it seems to be the same. Is this intentional?
(FWIW, I find the message in `related.py` to be clearer)